tls/s2n_connection.c

```diff
@@ -1708,3 +1710,137 @@ int s2n_connection_get_key_update_counts(struct s2n_connection *conn,
+        POSIX_GUARD(s2n_stuffer_write_bytes(&output, conn->secrets.version.tls13.resumption_master_secret, secret_size));
+    } else {
+        POSIX_GUARD(s2n_stuffer_write_bytes(&output, conn->secrets.version.tls12.master_secret, S2N_TLS_SECRET_LEN));
+        POSIX_GUARD(s2n_stuffer_write_bytes(&output, conn->handshake_params.client_random, S2N_TLS_RANDOM_DATA_LEN));
+        POSIX_GUARD(s2n_stuffer_write_bytes(&output, conn->handshake_params.server_random, S2N_TLS_RANDOM_DATA_LEN));
+    }
+
+    return S2N_SUCCESS;
+}
+
+int s2n_connection_set_traffic_context(struct s2n_connection *conn, uint8_t *buffer, uint32_t buffer_length)
+{
+    POSIX_ENSURE_REF(conn);
+    POSIX_ENSURE_REF(conn->secure);
+
+    struct s2n_blob context_blob = { 0 };
+    POSIX_GUARD(s2n_blob_init(&context_blob, buffer, buffer_length));
+    struct s2n_stuffer input = { 0 };
+    POSIX_GUARD(s2n_stuffer_init_written(&input, &context_blob));
+
+    uint64_t context_version = 0;
+    POSIX_GUARD(s2n_stuffer_read_uint64(&input, &context_version));
+    /* No other version is supported currently */
+    POSIX_ENSURE(context_version == S2N_TRAFFIC_CONTEXT_V1, S2N_ERR_INVALID_STATE);
+
+    POSIX_GUARD(s2n_stuffer_read_uint8(&input, &conn->actual_protocol_version));
+    conn->server_protocol_version = conn->actual_protocol_version;
+    conn->client_protocol_version = conn->actual_protocol_version;
+
+    uint8_t cipher_suite[2] = { 0 };
+    POSIX_GUARD(s2n_stuffer_read_bytes(&input, cipher_suite, S2N_TLS_CIPHER_SUITE_LEN));
+    POSIX_GUARD_RESULT(s2n_cipher_suite_from_iana(cipher_suite, S2N_TLS_CIPHER_SUITE_LEN, &conn->secure->cipher_suite));
+
+    /* Sequence numbers get zeroed during the TLS1.3 key expansion so they are read into temp variables */
+    uint8_t client_sequence_number[S2N_TLS_SEQUENCE_NUM_LEN] = { 0 };
+    uint8_t server_sequence_number[S2N_TLS_SEQUENCE_NUM_LEN] = { 0 };
+    POSIX_GUARD(s2n_stuffer_read_bytes(&input, client_sequence_number, S2N_TLS_SEQUENCE_NUM_LEN));
+    POSIX_GUARD(s2n_stuffer_read_bytes(&input, server_sequence_number, S2N_TLS_SEQUENCE_NUM_LEN));
+
+    POSIX_GUARD(s2n_stuffer_read_uint16(&input, &conn->max_outgoing_fragment_length));
+
+    if (conn->actual_protocol_version >= S2N_TLS13) {
+        uint8_t secret_size = 0;
+        POSIX_GUARD(s2n_hmac_digest_size(conn->secure->cipher_suite->prf_alg, &secret_size));
+
+        POSIX_GUARD(s2n_stuffer_read_bytes(&input, conn->secrets.version.tls13.client_app_secret, secret_size));
+        POSIX_GUARD(s2n_stuffer_read_bytes(&input, conn->secrets.version.tls13.server_app_secret, secret_size));
+        POSIX_GUARD(s2n_stuffer_read_bytes(&input, conn->secrets.version.tls13.resumption_master_secret, secret_size));
+
+        /* s2n_set_key checks that the handshake has completed, so here we artificially move through the state machine */
+        while (s2n_conn_get_current_message_type(conn) != APPLICATION_DATA) {
+            conn->handshake.message_number++;
+        }
```

Comment:
We do this a fair bit in tests, and have a testlib method for it: https://github.com/aws/s2n-tls/blob/main/tests/testlib/s2n_testlib.h#L78 Maybe we need a non-testlib method.